### PR TITLE
STOR-1692: CI implementation: Test the Vault provider with Secret Store CSI driver operator

### PR DESCRIPTION
**Pull request by ropatil010**
Work in progress, to check if this works fine with CI implementation.
/hold

### CODE REVIEW

**Comment by ropatil010** [comment]
Dependent PR: https://github.com/openshift/secrets-store-csi-driver/pull/16

**Comment by Phaow** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e

**Comment by ropatil010** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault

**Comment by ropatil010** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault

**Comment by Phaow** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault

File: ci-operator/step-registry/openshift/e2e/azure/csi/secrets-store/vault/openshift-e2e-azure-csi-secrets-store-vault-workflow.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    - chain: ipi-azure-pre
+    - ref: optional-operators-subscribe
+    - ref: storage-create-csi-secrets-store
+    - ref: storage-conf-wait-for-csi-driver
+    - ref: storage-obj-save
+    test:
+    - as: test-e2e-vault
+      cli: latest
+      commands: make e2e-vault
+      from: src
```
Comment (dobsonj) [inline ci-operator/step-registry/openshift/e2e/azure/csi/secrets-store/vault/openshift-e2e-azure-csi-secrets-store-vault-workflow.yaml:14]: > make: *** No rule to make target 'e2e-vault'.  Stop.

`from: src` refers to the operator src here, but this workflow needs the driver src to run [e2e-vault](https://github.com/openshift/secrets-store-csi-driver/blob/556773e6d44bc303fff28418f5e06760bcb80d47/Makefile#L481-L483).

You need to define a src image to be built by the driver, and then use that image here. I would try adding something like this to `ci-operator/config/openshift/secrets-store-csi-driver/openshift-secrets-store-csi-driver-main.yaml` in the `images` section:

```
- dockerfile_literal: |-
    FROM src
  from: src
  to: secrets-store-csi-driver-test
```

And then once we have that source image for the driver, your workflow can reference it with:

```
from: secrets-store-csi-driver-test
```

At least I think that will work... might take a little trial-and-error. (see also: [build-inputs docs](https://docs.ci.openshift.org/docs/architecture/ci-operator/#build-inputs))

**Comment by ropatil010** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault

File: ci-operator/config/openshift/secrets-store-csi-driver-operator/openshift-secrets-store-csi-driver-operator-main.yaml [hunk 2]
```diff
@@ -107,6 +107,19 @@ tests:
...
+    cluster_profile: azure4
+    dependencies:
+      OO_INDEX: ci-index-secrets-store-csi-driver-operator-bundle
+      SECRETS_STORE_E2E_PROVIDER_IMAGE: secrets-store-csi-driver-e2eprovider
+    env:
+      OO_CHANNEL: preview
+      OO_INSTALL_NAMESPACE: openshift-cluster-csi-drivers
+      OO_PACKAGE: secrets-store-csi-driver-operator
+      OO_TARGET_NAMESPACES: '!all'
+    workflow: openshift-e2e-azure-csi-secrets-store-vault
```
Comment (Phaow) [inline ci-operator/config/openshift/secrets-store-csi-driver-operator/openshift-secrets-store-csi-driver-operator-main.yaml:122]: Use the workflow, we need add the `secrets-store-csi-driver-testvault` image to base_images.
E.g.
https://github.com/openshift/release/pull/50084/files#diff-2c0a7b64192c771610e79d45c1ff783e4bae5ba5db161d0fc49de984920836afR30-R33

**Comment by ropatil010** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault

**Comment by ropatil010** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault

**Comment by ropatil010** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault

**Comment by ropatil010** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault

**Comment by ropatil010** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault

**Comment by dobsonj** [comment]
/retitle STOR-1692: CI implementation: Test the Vault provider with Secret Store CSI driver operator

**Comment by dobsonj** [comment]
The failures in the rehearsal are `sleep` related flakes from the test suite that I think we'll need to fix upstream. The CI-related changes in this PR though look good to me.

**Comment by dobsonj** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault pull-ci-openshift-secrets-store-csi-driver-operator-release-4.16-operator-e2e-vault pull-ci-openshift-secrets-store-csi-driver-operator-release-4.17-operator-e2e-vault

**Comment by ropatil010** [comment]
/unhold
@dobsonj @Phaow Guess we can merge this if no more corrections needed in this PR?

**Comment by ropatil010** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault pull-ci-openshift-secrets-store-csi-driver-operator-release-4.16-operator-e2e-vault pull-ci-openshift-secrets-store-csi-driver-operator-release-4.17-operator-e2e-vault

**Comment by ropatil010** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault

**Comment by ropatil010** [comment]
/pj-rehearse pull-ci-openshift-secrets-store-csi-driver-operator-main-operator-e2e-vault pull-ci-openshift-secrets-store-csi-driver-operator-release-4.16-operator-e2e-vault pull-ci-openshift-secrets-store-csi-driver-operator-release-4.17-operator-e2e-vault

**Comment by ropatil010** [comment]
Verification: 
Profile: ipi-on-azure/versioned-installer 
Payload: 4.16.0-0.nightly-2024-05-05-102537

Profile: ipi-on-azure/versioned-installer-sts 
Payload: 4.15.0-0.nightly-2024-05-04-030917 

rohitpatil@ropatil-mac secrets-store-csi-driver % make e2e-vault
bats -t test/bats/vault.bats
1..17
ok 1 install vault provider
ok 2 setup vault
ok 3 deploy vault secretproviderclass crd
ok 4 CSI inline volume test with pod portability
ok 5 CSI inline volume test with pod portability - read vault secret from pod
ok 6 CSI inline volume test with pod portability - rotation succeeds
ok 7 CSI inline volume test with pod portability - unmount succeeds
ok 8 Sync with K8s secrets - create deployment
ok 9 Sync with K8s secrets - read secret from pod, read K8s secret, read env var, check secret ownerReferences with multiple owners
ok 10 Sync with K8s secrets - delete deployment, check secret is deleted
ok 11 Test Namespaced scope SecretProviderClass - create deployment
ok 12 Test Namespaced scope SecretProviderClass - Sync with K8s secrets - read secret from pod, read K8s secret, read env var, check secret ownerReferences
ok 13 Test Namespaced scope SecretProviderClass - Sync with K8s secrets - delete deployment, check secret deleted
ok 14 Test Namespaced scope SecretProviderClass - Should fail when no secret provider class in same namespace
ok 15 deploy multiple vault secretproviderclass crd
ok 16 deploy pod with multiple secret provider class
ok 17 CSI inline volume test with multiple secret provider class

**Comment by dobsonj** [comment]
I don't like how flaky this job is, but we will have to fix it in the driver repo, and it would be useful to have this presubmit job in place for testing. So let's merge this and I'll follow up in https://issues.redhat.com/browse/STOR-1870 as part of moving the driver and operator to GA.
/lgtm
/approve
/pj-rehearse ack

**Pull request merged**